common/lib/xmodule/xmodule/modulestore/courseware_index.py

```diff
@@ -0,0 +1,118 @@
+""" Code to allow module store to interface with courseware index """
+from __future__ import absolute_import
+
+import logging
+
+from opaque_keys.edx.locator import CourseLocator
+from search.search_engine_base import SearchEngine
+
+from . import ModuleStoreEnum
+from .exceptions import ItemNotFoundError
+
+# Use default index and document names for now
+INDEX_NAME = "courseware_index"
+DOCUMENT_TYPE = "courseware_content"
+
+log = logging.getLogger('edx.modulestore')
+
+
+class IndexWriteError(Exception):
+
+    """ Raised to indicate that indexing of particular key failed """
+    pass
+
+
+class ModuleStoreCoursewareIndexMixin(object):
+
+    """
+    Mixin class to enable indexing for courseware search from different modulestores
+    """
+
+    def do_index(self, location, delete=False):
+        """
+        Main routine to index (for purposes of searching) from given location and other stuff on down
+        """
+        error = []
+        # TODO - inline for now, need to move this out to a celery task
+        searcher = SearchEngine.get_search_engine(INDEX_NAME)
+        if not searcher:
+            return
+
+        course_key = location if isinstance(location, CourseLocator) else location.course_key
+        location_info = {
+            "course": unicode(course_key),
+        }
+
+        def _fetch_item(item_location):
+            """ Fetch the item from the modulestore location, log if not found, but continue """
+            try:
+                if isinstance(item_location, CourseLocator):
+                    item = self.get_course(item_location)
+                else:
+                    item = self.get_item(item_location, revision=ModuleStoreEnum.RevisionOption.published_only)
+            except ItemNotFoundError:
+                log.warning('Cannot find: %s', item_location)
+                return None
+
+            return item
+
+        def index_item_location(item_location, current_start_date):
+            """ add this item to the search index """
+            item = _fetch_item(item_location)
+            if item:
+                if item.category in ['course', 'chapter', 'sequential', 'vertical', 'html', 'video']:
```

Comment:
@dino-cikatic We probably want to move this out into a setting... Although I'm still not sure why we need it here though - if we have additional index documents that have no content, then they won't show up in the search results anyway, and we won't have to remember to add new ones when we want their content.